User group detection + tweaks

StormySan opened the pull request:
Cleaning up usergroups - deleted users who had a group created could not then be added back, as shim didn't remove their group and attempted to re-create it.

Tweaking a few control flows to allow some older versions of Python to properly process them.

jamiesonbecker commented on shim.py:
```diff
@@ -89,17 +92,25 @@ def useradd(name, username, preferred_shell):
-    if os.path.isdir(home_dir):
-        useradd_suffix = ""
+    if not os.path.isdir(home_dir):
+        cmd.append(useradd_suffix = "-m")
```
Can you check this line?

jamiesonbecker commented:
Thanks for the pull request!! Looks great, and upgrading to a proper list for that cmd build out is a great idea, too. I'll check these switches for compatibility (weirdly, Debian-based, RHEL, and SUSE seem to sometimes have different command line switches, even for simple GNU utils like passwd or logger.)

jamiesonbecker commented:
Sorry, one more quick question with regard to your note. "deleted users who had a group created could not then be added back, as shim didn't remove their group and attempted to re-create it". We had a sort-of similar bug with user ID's a few months back. Under what circumstances did this happen; was it when one of their system created groups or a secondary group?

StormySan commented:
It was for users with a default group - the same as their name. Had a situation with a user who had been added to an openSUSE server, then deleted - the shim removed the user account, but didn't remove the group created for them (I added in lines to remove the group).

When the user was added back to the server, the shim tried to create the user and a group with their name, but as the group already existed the useradd command failed. I added a check to see if a group already exists with their name, and to add them to that if it exists - if not, then to create the group.

I unfortunately haven't had the opportunity to test it on other distros past openSUSE and CentOS, so it may be worth a check on some other versions.

jamiesonbecker commented:
Excellent, thanks Stormy! This is great stuff. We're going to be upgrading the front end and back end very soon (aiming for next week) and we'll get these changes integrated soon after. Leaving this open until then. (see note above, looks like invalid python?)

StormySan commented:
Ugh, sorry, I'll blame an unfortunate lack of caffeination on that slip-up!

jamiesonbecker commented:
100% understood :)

jamiesonbecker commented:
Getting ready for the migration and was able to reproduce these issues; almost ready to integrate. Thanks for submitting these Stormy.

StormySan commented:
My pleasure, happy to contribute :)

jamiesonbecker commented:
Hey Stormy, we're still putting together a CONTRIBUTORS doc, sorry for the delay!

jamiesonbecker commented:
Stormysan, trying to work through the legal things here (and greatly appreciate the pull req!) -- do you assign copyright for your work to Userify Corporation? Thanks!

StormySan commented:
Sure!

StormySan commented:
Haha sure thing - all work submitted in this pull request can be assigned to Userify Corp :)

jamiesonbecker commented:
Awesome, thank you!! :)

userifydevteam commented:
Unfortunately we should have moved forward on this in the last, oh, six years, but we didn't :( Sorry!

We'd welcome another pull request, but we're getting ready to release a completely new shim soon anyway. Stay tuned!
